### ISSUE TITLE
Implement dynamic assignments (where variable name isn't a constant)

### ISSUE THREAD

**Issue by Aversiste** [oilshell/oil#236]
Hello,

Trying one of my heavy shell project with OSH I stumbled upon this error: “Variable names must be unquoted constants”, caused by the following snippet:

```
echo test
a=foo
declare -ga $a
eval $a[0]="bar"
echo "${foo[0]}"
```

This is bash output:
```
test
bar
```

This script creates an indirection for the name of the variable foo which is created as a global array. I use this trick to create arrays in a portable manner despite array initialisation being one of the less portable feature of a modern shell.

The ksh equivalent syntax is not supported by OSH, but here it is for comparison:

    set -A $a -- bar

It works with mksh, OpenBSD ksh and zsh with a few setopt magic.

**Comment by andychu** [comment]
Great, thank you for the bug report.  I was hoping to avoid implementing this, but it's not surprising to me that someone is using it :)

I didn't know of this trick to use arrays portably across shells.  I think that bash, mksh, and ksh agree pretty well on arrays.  Is it zsh which is the outlier?

FWIW, I mentioned dynamic assignments here: http://www.oilshell.org/blog/2018/12/16.html#toc_4

**Comment by Aversiste** [comment]
Setting and accessing values in an array as well as measuring its length are effectively portable operations, it's only the initialization that is problematic as many different syntaxes exists.

zsh complicates things by providing nearly all these syntaxes but sometimes with subtle differences due to the huge option set. For example it is possible to declare arrays with `typeset -a` but also with `set -A` if the option ksharrays is activated.

mksh implements these too as well as bash style `name=(values)`. oksh doesn't.

These are the two functions I use regularly for this purpose, the first one working well with zsh, mksh, oksh and NetBSD's pdksh:

```
_arrayksh() {
        local _array_name="$1"; shift 
        unset "$_array_name"
        OLDIFS="$IFS"
        IFS="
"
        set -A $_array_name -- ${@:-''}
        IFS="$OLDIFS"
}

_arraybash() {
        local _array_name="$1"; shift
        unset $_array_name
        declare -ga $_array_name
        local _array_i=0
        local _array_j=0
        for _array_i in "$@"; do
                eval $_array_name[$_array_j]="\"$_array_i\""
                _array_j=$(( $_array_j + 1 ))
        done
}
```

**Comment by andychu** [comment]
NOTE: This will also let us re-unify `export` and `local/readonly` ....

**Comment by andychu** [comment]
Also reported in Zulip

**Comment by andychu** [comment]
Related crash reported in Zulip

```
$ osh -c 'dec=declare; $dec a=1; echo $a'
[python backtrace...]
NotImplementedError
```

**Comment by lhoursquentin** [comment]
Also noticed `alias` has no effect on assignment builtins:
```bash
osh$ alias declare=ls
osh$ declare /
  declare /
          ^
'/home/quentin/.config/oil/oshrc':2: Invalid variable name '/'
```

Technically aliases should be part of the "static" parsing right?

**Comment by andychu** [comment]
That is sort of mentioned here under alias expansion:

https://github.com/oilshell/oil/blob/master/doc/known-differences.md

(I just updated that doc today, and from now on it will published with every release under /release/$VERSION/.)

That paragraph could probably be made more precise / informative.  And I believe break/continue/return/exit exhibit the difference.

The way OSH works is that the whole file is statically parsed, and then "`SimpleCommand`" can be alias expanded.

For example, `ls -l` and `read -n 1` are `SimpleCommand`.  But `for x`, `declare x`, `{ ls ...`, `break`, `continue`, etc. aren't.

I would consider the difference to be a bug if some major shell script relied on some different behavior.  

i.e. filling out this list is how I judge the "importance" of differences:

https://github.com/oilshell/oil/wiki/Shell-Programs-That-Run-Under-OSH

-----

I mentioned the dynamic assignment issue on that page too, although I'm on the fence about it.  `bash-completion` uses dynamic assignment, which makes it important.  On the other hand, `bash-completion` won't run without (very minor) patches for some other fundamental reasons.

**Comment by andychu** [comment]
To summarize use cases for dynamic assignment:

- Aboriginal Linux used had a wrapper around export that did `export "$@"`.  Although I no longer think it's that important to run Aboriginal completely unmodified.  If a small  `eval` patch does it, that's OK.  OSH is more static by design.
- `bash-completion` uses it.  Again there are other reasons bash-completion will require small patches.
- The use case above from @Aversiste , which I wasn't familiar with.

And:

- It may fall out of refactoring to unify `a[$x]=$y` and `(( a[x] = y ))`.

So I want to do dynamic assignment eventually, but there are also a ton of things in front of it (e.g. performance and the Oil language)

**Comment by lhoursquentin** [comment]
Ok this makes more sense, thanks for the explanation!

> I would consider the difference to be a bug if some major shell script relied on some different behavior. But otherwise I wouldn't necessarily treat it as a bug.

Agreed. What bothers me is that neither `dec=declare; "$dec" var=1` nor `alias local=typeset` work. This is painful when trying to make a script as portable as possible. The workaround we talked about being to make use of `eval`, which is not ideal either.

For instance mksh has by default an alias to replace `local` by `typeset`

```bash
$ mksh -c alias | grep local
local='\\builtin typeset'
```

**Comment by andychu** [comment]
Released with 0.7.pre2: https://www.oilshell.org/release/0.7.pre2/
